### PR TITLE
Manual React setup

### PR DESCRIPTION
**Pull request by Jack-Music**
As discussed on discord, i tried to change te setup from create-react-app to a manual setup

- I would suggest to comment as many config options as we can, that every configuration is reasonable
- Maybe we should change .gitignore to allow vscode settings and recommendations (very common)
- It switches to four tabs as indent
- It keeps the service-worker and the icons

If you have any suggestions or corrections, please leave a comment.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "presets": [
+        "@babel/preset-env",
+        "@babel/preset-typescript",
+        "@babel/preset-react"
+    ],
+    "plugins": [
+        "@babel/proposal-class-properties",
+        "@babel/proposal-object-rest-spread"
+    ]
```
Comment (Udayraj123) [inline .babelrc:12]: These two can be good additions to this - 
```
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-nullish-coalescing-operator"
```

File: .eslintignore [hunk 2]
```diff
@@ -1 +1,2 @@
-public/
\ No newline at end of file
+public/
+webpack.config.js
```
Comment (Udayraj123) [inline .eslintignore:2]: Why should `webpack.config.js` be ignored? How can other developers build it locally?

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        "plugin:@typescript-eslint/eslint-recommended",
+        "plugin:@typescript-eslint/recommended",
+        "plugin:prettier/recommended",
+        "prettier/@typescript-eslint"
+    ],
+    "parserOptions": {
+        "ecmaVersion": 2020,
+        "sourceType": "module"
+    },
+    "ecmaFeatures": { "jsx": true },
```
Comment (Udayraj123) [inline .eslintrc]: I think ecmaFeatures has to be inside parserOptions.
https://eslint.org/docs/user-guide/configuring

File: package.json [hunk 4]
```diff
@@ -1,57 +1,75 @@
...
-	},
-	"devDependencies": {
-		"husky": "^4.2.5",
-		"lint-staged": "^10.2.6",
-		"prettier": "^2.0.5"
-	}
+    "name": "virtual-mmx",
+    "version": "0.1.0-alpha",
+    "private": true,
+    "description": "Web application simulating Wintergatans Marble Maschine X",
```
Comment (Udayraj123) [inline package.json]: Typo with Maschine

File: package.json [hunk 5]
```diff
@@ -1,57 +1,75 @@
...
+        "babel-loader": "^8.1.0",
+        "css-loader": "^3.5.3",
+        "eslint": "^7.1.0",
+        "eslint-config-prettier": "^6.11.0",
+        "eslint-plugin-prettier": "^3.1.3",
+        "eslint-plugin-react": "^7.20.0",
+        "html-webpack-plugin": "^4.3.0",
+        "style-loader": "^1.2.1",
+        "typescript": "^3.9.3",
+        "webpack": "^4.43.0",
```
Comment (Udayraj123) [inline package.json:43]: Just a thought, since we've started with this config - why not use webpack 5?

File: public/manifest.json [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
-			"type": "image/png",
-			"sizes": "512x512"
-		}
-	],
-	"start_url": ".",
-	"display": "standalone",
-	"theme_color": "#000000",
-	"background_color": "#ffffff"
+    "short_name": "vMMX",
+    "name": "virtual Marbe Machine X",
```
Comment (Udayraj123) [inline public/manifest.json]: typo with Marbe

File: tsconfig.json [hunk 7]
```diff
@@ -1,5 +1,10 @@
 {
+	"compileOnSave": false,
 	"compilerOptions": {
+		"noFallthroughCasesInSwitch": true,
+		"noImplicitAny": true,
+		"noUnusedLocals": true,
+		"outDir": "./dist/",
 		"target": "es5",
 		"lib": ["dom", "dom.iterable", "esnext"],
 		"allowJs": true,
```
Comment (Udayraj123) [inline tsconfig.json]: Additionally use this - 
```
  "exclude": ["node_modules"]
```

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
+const path = require('path');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+module.exports = {
+    // webpack will take the files from ./src/index
+    entry: './src/index',
```
Comment (Udayraj123) [inline webpack.config.js:5]: does it work without specifying the extension?

File: webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    // and output it into /dist as bundle.js
+    output: {
+        path: path.join(__dirname, '/dist'),
+        filename: 'bundle.js',
```
Comment (Udayraj123) [inline webpack.config.js]: Probably better to use `filename : 'bundle.[hash].js'` so that user won't have to hard reload for changes to take effect

**Review by Udayraj123 — CHANGES_REQUESTED** [review]

**Comment by Udayraj123** [comment]
Overall this stuff shall improve the collaborative developer experience by a good magnitude 😄

File: package.json [hunk 10]
```diff
@@ -1,57 +1,75 @@
...
+        "webpack-cli": "^3.3.11",
+        "webpack-dev-server": "^3.11.0"
+    },
+    "scripts": {
+        "bundle": "webpack",
+        "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (Udayraj123) [inline package.json]: Make it `exit 0` for now to pass the checks

File: package.json [hunk 11]
```diff
@@ -1,57 +1,75 @@
...
+        "check-types": "tsc"
+    },
+    "husky": {
+        "hooks": {
+            "pre-commit": "lint-staged"
+        }
+    },
+    "lint-staged": {
+        "*.{js,ts,tsx}": "eslint --cache --fix --ignore-path .eslintignore",
+        "*.{css,md}": "prettier --write --ignore-path .eslintignore"
```
Comment (Udayraj123) [inline package.json:61]: prettier should apply to js/jsx/tsx too- 
```
	"lint-staged": {
		"*.{js,ts,tsx}": "eslint --cache --fix --ignore-path .eslintignore",
		"*.{js,ts,tsx,css,md}": "prettier --write --ignore-path .eslintignore"
	},
```

File: public/index.html [hunk 12]
```diff
@@ -1,35 +1,24 @@
...
-      work correctly both with client-side routing and a non-root public URL.
-      Learn how to configure a non-root public URL by running `npm run build`.
-    -->
-		<title>React App</title>
-	</head>
-	<body>
-		<noscript>You need to enable JavaScript to run this app.</noscript>
-		<div id="root"></div>
-		<!--
+        <title>React App</title>
```
Comment (FelixWohlfrom) [inline public/index.html]: Do we really want to keep this title?

File: .eslintignore (hunk 2 above)
Comment (Jack-Music) [reply .eslintignore:2]: I just thought `wepack.config.js` should not be linted

File: tsconfig.json (hunk 7 above)
Comment (Jack-Music) [reply tsconfig.json]: I will add it, but is it really necessary when using: `"include": ["src"]`?

File: package.json (hunk 5 above)
Comment (Jack-Music) [reply package.json:43]: As far as i know, webpack 5.0 is still in beta.
https://github.com/webpack/webpack/releases

File: .babelrc (hunk 1 above)
Comment (Jack-Music) [reply .babelrc:12]: Added in 6e66da9d377e7dbd1846ab86eb016da26ef1b156

File: .eslintrc (hunk 3 above)
Comment (Jack-Music) [reply .eslintrc]: Thanks, fixed in 6e66da9d377e7dbd1846ab86eb016da26ef1b156

File: package.json (hunk 4 above)
Comment (Jack-Music) [reply package.json]: Fixed in 6e66da9d377e7dbd1846ab86eb016da26ef1b156

File: public/index.html (hunk 12 above)
Comment (Jack-Music) [reply public/index.html]: Thank you for the review!
I replaced it with vMMX in 6e66da9d377e7dbd1846ab86eb016da26ef1b156.
Do you have suggestions for a better title?

File: package.json (hunk 10 above)
Comment (Jack-Music) [reply package.json]: Fixed in 6e66da9d377e7dbd1846ab86eb016da26ef1b156

File: package.json (hunk 11 above)
Comment (Jack-Music) [reply package.json:61]: This plugin `"plugin:prettier/recommended"` in `.eslintrc` starts prettier with eslint.
So prettier indirectly applies to js/jsx/tsx.

File: webpack.config.js (hunk 9 above)
Comment (Jack-Music) [reply webpack.config.js]: Changed in 6e66da9d377e7dbd1846ab86eb016da26ef1b156

File: public/index.html (hunk 12 above)
Comment (FelixWohlfrom) [reply public/index.html]: vMMX sounds good for me - maybe we can later extend it with the currently loaded song.

File: package.json (hunk 11 above)
Comment (Udayraj123) [reply package.json:61]: Okay, wasn't aware of that feat. Cool!

File: tsconfig.json (hunk 7 above)
Comment (Udayraj123) [reply tsconfig.json]: Yes. No harm in explicitly ignoring node_modules I think

File: .eslintignore (hunk 2 above)
Comment (Udayraj123) [reply .eslintignore:2]: Ah, my bad. Read the filename as gitignore

File: public/index.html (hunk 12 above)
Comment (Udayraj123) [reply public/index.html]: Let's now add a favicon too now that we have a new logo 😄

Comment (Udayraj123) [reply public/index.html]: Okay nvm it's already added

**Comment by Udayraj123** [comment]
Merged into dev branch https://github.com/wintergatan-community/virtual-mmx/pull/31

**Comment by mozi-h** [comment]
If this PR is no longer nececcary, you can close it, @Udayraj123 
(Note: This has a more elaborate description and a longer history than your new PR, so you might want to continue #30's work here?)

**Comment by Udayraj123** [comment]
@mozi-h #30 is done too. And I can close this now. The history will be preserved well given we've mentioned this issue there too.

**Pull request closed without merging**